batch2/batch/batch.py

```diff
@@ -1,19 +1,1278 @@
+    return jsonify(await _get_batches_list(params, user))
+
+
+@routes.post('/api/v1alpha/batches/{batch_id}/jobs/create')
+@prom_async_time(REQUEST_TIME_POST_CREATE_JOBS)
+@rest_authenticated_users_only
+async def create_jobs(request, userdata):
+    start = time.time()
+    batch_id = int(request.match_info['batch_id'])
+    user = userdata['username']
+
+    start1 = time.time()
+    batch = await Batch.from_db(batch_id, user)
+    log.info(f'took {round(time.time() - start1, 3)} seconds to get batch from db')
+
+    if not batch:
+        abort(404)
+    if batch.closed:
+        abort(400, f'batch {batch_id} is already closed')
+
+    start2 = time.time()
+    jobs_parameters = await request.json()
+    log.info(f'took {round(time.time() - start2, 3)} seconds to get data from server')
+
+    start3 = time.time()
+    validator = cerberus.Validator(schemas.job_array_schema)
+    if not await blocking_to_async(app['blocking_pool'], validator.validate, jobs_parameters):
+        abort(400, 'invalid request: {}'.format(validator.errors))
+    log.info(f"took {round(time.time() - start3, 3)} seconds to validate spec")
+
+    start4 = time.time()
+    jobs_builder = JobsBuilder(db)
+    try:
+        for job_params in jobs_parameters['jobs']:
+            create_job(jobs_builder, batch.id, userdata, job_params)
+
+        success = await jobs_builder.commit()
+        if not success:
+            abort(400, f'insertion of jobs in db failed')
+    finally:
+        await jobs_builder.close()
+
+    log.info(f'took {round(time.time() - start4, 3)} seconds to commit jobs to db')
+
+    log.info(f'took {round(time.time() - start, 3)} seconds to create jobs from start to finish')
+    return jsonify({})
+
+
+@routes.post('/api/v1alpha/batches/create')
+@prom_async_time(REQUEST_TIME_POST_CREATE_BATCH)
+@rest_authenticated_users_only
+async def create_batch(request, userdata):
+    start = time.time()
+    parameters = await request.json()
+
+    validator = cerberus.Validator(schemas.batch_schema)
+    if not validator.validate(parameters):
+        abort(400, 'invalid request: {}'.format(validator.errors))
+
+    batch = await Batch.create_batch(
+        attributes=parameters.get('attributes'),
+        callback=parameters.get('callback'),
+        userdata=userdata,
+        n_jobs=parameters['n_jobs'])
+    if batch is None:
+        abort(400, f'creation of batch in db failed')
+
+    log.info(f'took {round(time.time() - start, 3)} seconds to initialize batch {batch.id} in db')
+    return jsonify(await batch.to_dict(include_jobs=False))
+
+
+async def _get_batch(batch_id, user, limit=None, offset=None):
+    batch = await Batch.from_db(batch_id, user)
+    if not batch:
+        abort(404)
+    return await batch.to_dict(include_jobs=True, limit=limit, offset=offset)
+
+
+async def _cancel_batch(batch_id, user):
+    batch = await Batch.from_db(batch_id, user)
+    if not batch:
+        abort(404)
+    asyncio.ensure_future(batch.cancel())
```

Comment:
batch.cancel already uses an ensure_future to run the cancel in the background. let's not use it here. In general, I think we need to parsimoniously use ensure_future because it requires us to reason about concurrent code.